crates/pathfinder/src/state/merkle_tree.rs

```diff
@@ -1526,4 +1532,28 @@ mod tests {
             );
         }
     }
+
+    #[test]
+    fn dfs_on_leaf_to_binary_collision_tree() {
+        let mut conn = rusqlite::Connection::open_in_memory().unwrap();
+        let transaction = conn.transaction().unwrap();
+        let mut uut = MerkleTree::load("test", &transaction, StarkHash::ZERO).unwrap();
+
+        let value = starkhash!("01");
+        let key0 = starkhash!("ee00").view_bits().to_bitvec();
+        let key1 = starkhash!("ee01").view_bits().to_bitvec();
+
+        let key2 = starkhash!("ffff").view_bits().to_bitvec();
+        let hash_of_values = stark_hash::stark_hash(value, value);
+        uut.set(&key2, hash_of_values).unwrap();
+
+        uut.set(&key0, value).unwrap();
+        uut.set(&key1, value).unwrap();
+
+        let root = uut.commit().unwrap();
+
+        let uut = MerkleTree::load("test", &transaction, root).unwrap();
+        // this used to panic because it did find the binary on dev profile with the leaf hash
+        uut.dfs(&mut |_| {});
+    }
```

Comment:
Maybe worth adding the assertions, too?
```suggestion
            assert_eq!(uut.get(&key0).unwrap(), value);
            assert_eq!(uut.get(&key1).unwrap(), value);
            assert_eq!(uut.get(&key2).unwrap(), hash_of_values);
    }
```